system/jlib/jtrace.hpp

```diff
@@ -0,0 +1,66 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+
+
+#ifndef JTRACE_HPP
+#define JTRACE_HPP
+
+enum class TraceFlags : unsigned
+{
+    None = 0,
+    Standard = 1,
+    Detailed = 2,
+    Max = 3,
+    LevelMask = 0x3,
+    // Individual bit values above low two bits may be given meanings by specific engines or contexts, using constexpr to rename the following
+    All = 0,         // Use with doTrace to test JUST the level, without a feature flag
+    flag1 = 0x4,
+    flag2 = 0x8,
+    flag3 = 0x10,
+    flag4 = 0x20,
+    flag5 = 0x40,
+    flag6 = 0x80,
+    flag7 = 0x100,
+    flag8 = 0x200,
+    flag9 = 0x400,
+    flag10 = 0x800,
+    flag11 = 0x1000,
+    flag12 = 0x2000,
+    flag13 = 0x4000,
+    flag14 = 0x8000,
+};
+BITMASK_ENUM(TraceFlags);
+
```

Comment:
Some comments.

I think the idea broadly works, but I would change it in the following ways:

- include all the flags for all the components in this file.
- use a different name for the jlib flags (and any other very widely-shared library e.g. jhtree) - so that they can be added and flags renumbered without causing grief.

Introduce a helper function to make it easy to define and process the options for a component.  Something along the lines of....
```

struct TraceOption { const char * name; unsigned value; };
#define OPT(NAME) { "@" # NAME, TraceFlag::NAME }
TraceFlag getTraceFlags(IConstWorkunit * wu, const std::initializer_list<TraceOption> & y, TraceFlag dft);
TraceFlag getTraceFlags(IPropertyTree * globals, const std::initializer_list<TraceOption> & y, TraceFlag dft);
-----
constexpr std::initializer_list<TraceOption> roxieOptions{ OPT(traceX), OPT(traceY)) };
roxieTraceFlags = getTraceFlags(globals, roxieOptions, TraceFlag::none);
roxieTraceFlags = getTraceFlags(wu, roxieOptions, roxieTraceFlags);
```

There would cover the vast majority of cases.  There are two cases that may need different processing, where the category may be more fine grained.
- tracing on per activity type (in thor/roxie?)
- tracing on a per transform type (in code generator)

I'm not sure if that granularity is need in thor.  It would be useful in the code generator, but it looks like there are n the code generator I'm not sure how many different types if transform there are 107 different transformations so they wouldn't fit into that scheme.  Probably something similar using an array of booleans would work. It can be within the code generator - but any other component with similar issues should use the same approach, so the helper functions should be global.